client/state/data-layer/wpcom/comments/index.js

```diff
@@ -156,6 +160,35 @@ export const announceFailure = ( { dispatch, getState }, { siteId, postId } ) =>
 	dispatch( errorNotice( error ) );
 };
 
+export const deleteComment = ( { dispatch }, action ) => {
+	const { siteId, commentId } = action;
+	dispatch(
+		http( {
+			method: 'POST',
+			apiVersion: '1.1',
+			path: `/sites/${ siteId }/comments/${ commentId }/delete`,
+			onSuccess: {
+				...action,
+				type: COMMENTS_DELETE_SUCCESS,
+			},
+			onFailure: {
+				...action,
+				type: COMMENTS_DELETE_FAILURE,
+			},
+		} )
+	);
+};
+
+export const removeCommentFromState = ( { dispatch }, action ) => {
+	const { siteId, postId, commentId } = action;
+	dispatch( removeComment( siteId, postId, commentId ) );
+};
+
+export const announceDeleteFailure = ( { dispatch } ) => {
+	dispatch( errorNotice( translate( 'Could not delete the comment' ) ) );
+};
```

Comment:
> If we only remove the comment from the state onSuccess, why would we need to re-add it onFailure?

ah. see the reducers are responsible for deleting a comment from Calypso whenever they see `COMMENTS_DELETE`. It's a done-deal as far as they're concerned. Who cares if there's an API call. The data layer makes sure that WordPress.com is aware that the comment was deleted.

Therefore, `onSuccess` we have nothing more to do actually. It's only when the API request to actually delete the comment fails that we need to respond. Doing nothing but announce is probably good enough; the comment will return when they refresh their list of comments. On the other hand, if we add that comment back into state with `COMMENTS_ADD` then it also reappears because the delete failed.

> Also: not sure what local does though. Is it a way to skip the data layer and just do stuff into the Redux state?

yes. I added this to the original comment right after I wrote it.

_all_ actions pass through the data layer to the reducers (except network call responses) so they will see the actions before the API calls finish. if we want to add a comment but we don't want a new API request to create one, we have to wrap in `local`. it will be renamed eventually to `bypassDataLayer()`